### PR TITLE
Fix the parsing of Sarif files when the path is relative

### LINKED ISSUE #4100: C# analysis failing hard when source generators are used
Reported on the community forum: https://community.sonarsource.com/t/c-analysis-failing-in-azure-pipelines-due-to-java-lang-illegalargumentexception-uri-is-not-absolute/38995 . Project reproducer, logs and all information is present there.

> The .CommandGenerator.cs files are indeed generated during build time. This is triggered by using the [GenerateCommand] attribute in WpfDemo/DemoViewModel.cs.

The SARIF report parsing is failing hard:

```
2021-02-23T21:59:23.4779322Z WARN: File 'Generators\Generators.CommandGenerator\GenerateCommandAttribute.cs' referenced by the protobuf 'MetricsInfo' does not exist in the analysis context
2021-02-23T21:59:23.5134921Z WARN: File 'Generators\Generators.CommandGenerator\DemoViewModel.CommandGenerator.cs' referenced by the protobuf 'MetricsInfo' does not exist in the analysis context
2021-02-23T21:59:23.5136692Z WARN: File 'Generators\Generators.CommandGenerator\GenerateCommandAttribute.cs' referenced by the protobuf 'TokenTypeInfo' does not exist in the analysis context
2021-02-23T21:59:23.5138016Z WARN: File 'Generators\Generators.CommandGenerator\DemoViewModel.CommandGenerator.cs' referenced by the protobuf 'TokenTypeInfo' does not exist in the analysis context
2021-02-23T21:59:23.5139426Z WARN: File 'Generators\Generators.CommandGenerator\GenerateCommandAttribute.cs' referenced by the protobuf 'SymbolReferenceInfo' does not exist in the analysis context
2021-02-23T21:59:23.5142730Z WARN: File 'Generators\Generators.CommandGenerator\DemoViewModel.CommandGenerator.cs' referenced by the protobuf 'SymbolReferenceInfo' does not exist in the analysis context
2021-02-23T21:59:23.5144461Z WARN: File 'Generators\Generators.CommandGenerator\GenerateCommandAttribute.cs' referenced by the protobuf 'CopyPasteTokenInfo' does not exist in the analysis context
2021-02-23T21:59:23.5145858Z WARN: File 'Generators\Generators.CommandGenerator\DemoViewModel.CommandGenerator.cs' referenced by the protobuf 'CopyPasteTokenInfo' does not exist in the analysis context
2021-02-23T21:59:23.5240639Z INFO: Importing 2 Roslyn reports
2021-02-23T21:59:23.8892188Z INFO: ------------------------------------------------------------------------
2021-02-23T21:59:23.9082600Z INFO: EXECUTION FAILURE
2021-02-23T21:59:23.9099757Z INFO: ------------------------------------------------------------------------
2021-02-23T21:59:23.9103004Z INFO: Total time: 43.754s
2021-02-23T21:59:23.9116543Z INFO: Final Memory: 27M/90M
2021-02-23T21:59:23.9118045Z INFO: ------------------------------------------------------------------------
2021-02-23T21:59:23.9119856Z ##[error]ERROR: Error during SonarScanner execution
2021-02-23T21:59:23.9121605Z ERROR: Error during SonarScanner execution
2021-02-23T21:59:23.9123175Z ##[error]java.lang.IllegalArgumentException: URI is not absolute
2021-02-23T21:59:23.9128841Z java.lang.IllegalArgumentException: URI is not absolute
2021-02-23T21:59:23.9130496Z ##[error]at java.base/java.io.File.<init>(File.java:418)
2021-02-23T21:59:23.9134877Z 	at java.base/java.io.File.<init>(File.java:418)
2021-02-23T21:59:23.9136986Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.uriToAbsolutePath(SarifParser10.java:225)
2021-02-23T21:59:23.9139038Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.uriToAbsolutePath(SarifParser10.java:225)
2021-02-23T21:59:23.9140788Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleLocation(SarifParser10.java:175)
2021-02-23T21:59:23.9142659Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleLocation(SarifParser10.java:175)
2021-02-23T21:59:23.9144410Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleResultFileElement(SarifParser10.java:146)
2021-02-23T21:59:23.9146191Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleResultFileElement(SarifParser10.java:146)
2021-02-23T21:59:23.9147940Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleLocationsElement(SarifParser10.java:138)
2021-02-23T21:59:23.9149709Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleLocationsElement(SarifParser10.java:138)
2021-02-23T21:59:23.9151397Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleIssue(SarifParser10.java:107)
2021-02-23T21:59:23.9153212Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleIssue(SarifParser10.java:107)
2021-02-23T21:59:23.9155243Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleIssues(SarifParser10.java:95)
2021-02-23T21:59:23.9157486Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleIssues(SarifParser10.java:95)
2021-02-23T21:59:23.9159270Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.accept(SarifParser10.java:65)
2021-02-23T21:59:23.9161073Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.accept(SarifParser10.java:65)
2021-02-23T21:59:23.9162930Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.accept(SarifParser10.java:39)
2021-02-23T21:59:23.9164756Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.accept(SarifParser10.java:39)
2021-02-23T21:59:23.9166545Z ##[error]at org.sonarsource.dotnet.shared.plugins.RoslynDataImporter.importRoslynReports(RoslynDataImporter.java:54)
2021-02-23T21:59:23.9168423Z 	at org.sonarsource.dotnet.shared.plugins.RoslynDataImporter.importRoslynReports(RoslynDataImporter.java:54)
2021-02-23T21:59:23.9170358Z ##[error]at org.sonarsource.dotnet.shared.plugins.DotNetSensor.executeInternal(DotNetSensor.java:111)
2021-02-23T21:59:23.9173852Z 	at org.sonarsource.dotnet.shared.plugins.DotNetSensor.executeInternal(DotNetSensor.java:111)
2021-02-23T21:59:23.9175845Z ##[error]at org.sonarsource.dotnet.shared.plugins.DotNetSensor.execute(DotNetSensor.java:71)
2021-02-23T21:59:23.9177777Z 	at org.sonarsource.dotnet.shared.plugins.DotNetSensor.execute(DotNetSensor.java:71)
```

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes #4100

### CODE REVIEW

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/sarif/SarifParser10Test.java [hunk 1]
```diff
@@ -226,4 +226,27 @@ public void sarif_version_1_0_secondary_locations_messages() throws IOException
...
+
+    Location s1144PrimaryLocation = new Location("SourceGeneratorPOC.Generators\\SourceGeneratorPOC.SourceGenerator\\Greetings.cs", "Remove the unused private method 'UnusedMethod'.", 7, 8, 7, 46);
+    inOrder.verify(callback).onIssue("S1144", "warning", s1144PrimaryLocation, new ArrayList<>());
+
+    Location s1186PrimaryLocation = new Location("SourceGeneratorPOC.Generators\\SourceGeneratorPOC.SourceGenerator\\Greetings.cs", "Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation.", 7, 28, 7, 40);
+    inOrder.verify(callback).onIssue("S1186", "warning", s1186PrimaryLocation, new ArrayList<>());
+  }
+
+  @Test
+  public void sarif_version_1_0_region_with_length() throws IOException {
```
Comment (costin-zaharia-sonarsource) [inline sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/sarif/SarifParser10Test.java:243]: This case, although supported was not covered by unit tests. 

While working on this I've done also a double check to see if the `length` it's documented as a part of the `region` properties in the [json schema](http://json.schemastore.org/sarif-1.0.0) and it is.

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

**Comment by andrei-epure-sonarsource** [comment]
I realized that for #4100 we don't have an integration test... @costin-zaharia-sonarsource - what's the behavior for source generators - the file isn't indexed, so it's not reported, right? Should we have an IT?

**Comment by costin-zaharia-sonarsource** [comment]
> what's the behavior for source generators - the file isn't indexed, so it's not reported, right?

Yes, we will not import issues if the files are missing.

> Should we have an IT?

I think it would be helpful and I cannot find a test which covers the case when we have issues but not sources. I will add one.

**Pull request merged**